### PR TITLE
layerscape: fix problems with sysupgrade, default packages and boot properties 

### PR DESCRIPTION
**Pull request by CHKDSK88**
List of improvements:
- [ ] Fix sysupgrade with sd-card images
- [x] Fix LS1012-FRDM booting
- [x] Add missing default packages
- [ ] Add dtb to sysupgrade
- [ ] Make sd-card images combined

### CODE REVIEW

File: target/linux/layerscape/image/armv8_64b.mk [hunk 1]
```diff
@@ -102,7 +102,8 @@ define Device/fsl_ls1043a-rdb
   DEVICE_PACKAGES += \
     layerscape-fman \
     tfa-ls1043a-rdb \
-    fmc fmc-eth-config
+    fmc fmc-eth-config \
+    kmod-hwmon-ina2xx kmod-hwmon-lm90
```
Comment (adschm) [inline target/linux/layerscape/image/armv8_64b.mk]: This Makefile seems to have a one-package-per-line scheme, please maintain that for your additions.

File: target/linux/layerscape/base-files/lib/upgrade/platform.sh [hunk 2]
```diff
@@ -107,15 +107,29 @@ platform_do_upgrade_traverse_nandubi() {
...
+		mount -t $parttype -o rw,noatime "/dev/$partdev" /mnt 2>&1
 		cp -af "$UPGRADE_BACKUP" "/mnt/$BACKUP_FILE"
 		umount /mnt
 	fi
 }
+platform_copy_config() {
+	local board=$(board_name)
+
+	case "$board" in
+	*sdboot*)
```
Comment (adschm) [inline target/linux/layerscape/base-files/lib/upgrade/platform.sh:71]: I'd prefer to list the devices explicitly, as using wildcards makes it impossible to grep for board names.

File: target/linux/layerscape/base-files/lib/preinit/79_move_config [hunk 3]
```diff
@@ -0,0 +1,21 @@
+. /lib/functions.sh
+. /lib/upgrade/common.sh
+
+BOOTPART=/dev/mmcblk0p1
+
+move_config() {
+	local board=$(board_name)
+
+	case "$board" in
+	*sdboot*)
```
Comment (adschm) [inline target/linux/layerscape/base-files/lib/preinit/79_move_config:10]: Same as above, I'd prefer explicit names.

File: target/linux/layerscape/base-files/lib/upgrade/platform.sh [hunk 4]
```diff
@@ -140,22 +83,19 @@ platform_check_image() {
...
 	fsl,ls1012a-rdb | \
 	fsl,ls1021a-twr | \
+	fsl,ls1021a-twr-sdboot | \
 	fsl,ls1043a-rdb | \
+	fsl,ls1043a-rdb-sdboot | \
 	fsl,ls1046a-rdb | \
+	fsl,ls1046a-rdb-sdboot | \
 	fsl,ls1088a-rdb | \
+	fsl,ls1088a-rdb-sdboot | \
 	fsl,ls2088a-rdb)
```
[the remark this answers is not in the document]
Comment (adschm) [reply target/linux/layerscape/base-files/lib/upgrade/platform.sh:96]: One argument speaking against it would be that logic would be different from platform_do_upgrade then ...

File: target/linux/layerscape/image/armv7.mk [hunk 5]
```diff
@@ -29,6 +29,12 @@ define Device/fix-sysupgrade
 	After that, You can use "sysupgrade -F".
 endef
 
+define Device/rework-sdcard-images
+  DEVICE_COMPAT_VERSION := 2.0
+  DEVICE_COMPAT_MESSAGE := SD-card images was reworked to squashfs + ext4 overlay combined images. \
+	It is required to flash entire sd-card again and manually copy config.
+endef
```
Comment (adschm) [inline target/linux/layerscape/image/armv7.mk:36]: If used in both files, I'd move this to the target's image/Makefile.

File: target/linux/layerscape/image/armv8_64b.mk [hunk 6]
```diff
@@ -138,7 +138,7 @@ define Device/fsl_ls1043a-rdb-sdboot
     layerscape-fman \
     tfa-ls1043a-rdb-sdboot \
     fmc fmc-eth-config \
-    kmod-hwmon-ina2xx kmod-hwmon-lm90
+    kmod-ahci-qoriq kmod-hwmon-ina2xx kmod-hwmon-lm90
```
Comment (adschm) [inline target/linux/layerscape/image/armv8_64b.mk:141]: One line per package, see above.

**Comment by adschm** [comment]
I've picked the first patch already as it's a no-brainer.

File: target/linux/layerscape/image/armv8_64b.mk (hunk 1 above)
Comment (CHKDSK88) [reply target/linux/layerscape/image/armv8_64b.mk]: It is mixed. Some packages with the same first letter are together:  ```fmc fmc-eth-config``` or  ```kmod-gpio-pca953x kmod-input-gpio-keys-polled \```. Some not.
But I will change it, if you want.

Comment (adschm) [reply target/linux/layerscape/image/armv8_64b.mk]: I see what you mean. However, it appears to me that the original idea was to keep one line per packages, and I'd stick to that for adding new stuff (actually, that scheme is more convenient for some use-cases).

The "fmc fmc-eth-config" is at least consistently deviating from that, and I just would ignore it. The "kmod-gpio-pca953x kmod-input-gpio-keys-polled" actually should be converted into two lines, but I don't think it would be worth a commit just for that.

**Comment by adschm** [comment]
So, we got the first use-case for major compat-version increase ... Much earlier than I thought.

**Comment by CHKDSK88** [comment]
@adschm 
> So, we got the first use-case for major compat-version increase ... Much earlier than I thought.

And I found first issue. I'm not sure if it is problem with my code or main repo. At this moment every upgrade causes 1.0->2.0 error. Even when I try flash factory 2.0 image with 2.0 sysupgrade image.

**Comment by adschm** [comment]
> @adschm
> 
> > So, we got the first use-case for major compat-version increase ... Much earlier than I thought.
> 
> And I found first issue. I'm not sure if it is problem with my code or main repo. At this moment every upgrade causes 1.0->2.0 error. Even when I try flash factory 2.0 image with 2.0 sysupgrade image.

Oh, I actually didn't think it through when looking for the first time:

The DEVICE_COMPAT_VERSION will only affect the _image_, i.e. the supported_devices metadata, but not write anything into the 
"content" by itself. 

So, you additionally need to write the "on-device" value to uci config, as e.g. done here:
https://github.com/openwrt/openwrt/commit/090779b24e436bb5113719eb7067c7b8151abd14

As in this kirkwood example, both Makefile and board.d changes are necessary.

However, in your case the situation is even different: The way I update compat-version for the config as done in board.d file above will only work when board.json is recreated, i.e. when config is wiped beforehand. As I understand your situation, you just want to prompt the user to update bootargs, but would keep config during sysupgrade. Then, the board.d approach won't work, but you could use uci-defaults as done here: https://github.com/openwrt/openwrt/commit/86c89bf5e8f570214371824587b38220cf05bdac

For kirkwood, this is just a workaround for early adopters, as users actually _should_ wipe their config. In your case, though, it might make more sense to use the uci-defaults solution, as this would overwrite existing config. Note that this would also overwrite compat-version in config if the user forced without applying the requested change, but is no issue in your case, as then the device won't boot anyhow ...

Long story short, you should just copy the uci-defaults/05_fix-compat-version file from the latest link and add the relevant devices to it. This makes most sense for the dtb patch. For the sdcard patch, both concepts would work, as they user would get default config here.

File: target/linux/layerscape/base-files/lib/upgrade/platform.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (CHKDSK88) [reply target/linux/layerscape/base-files/lib/upgrade/platform.sh:100]: At this moment I prefer keep as is.

**Comment by CHKDSK88** [comment]
@adschm 

All changes done. 

I rebased to current master. It causes next patch contains fix one of layerscape dsa patches.

**Comment by CHKDSK88** [comment]
Rebased to current master and drop patch with merged fix.

File: target/linux/layerscape/image/armv7.mk [hunk 7]
```diff
@@ -16,12 +16,14 @@ define Device/Default
   DEVICE_DTS = $(lastword $(subst _, ,$(1)))
   SUPPORTED_DEVICES = $(subst _,$(comma),$(1))
   IMAGE_SIZE := 64m
-  IMAGE/sysupgrade.bin := append-kernel | pad-to 16M | \
+  IMAGE/sysupgrade.bin = ls-append-dtb $$(DEVICE_DTS) | pad-to 1M | \
+	append-kernel | pad-to 17M | \
```
Comment (adschm) [inline target/linux/layerscape/image/armv7.mk:20]: Indent should be spaces here.

File: target/linux/layerscape/image/armv8_64b.mk [hunk 8]
```diff
@@ -15,7 +15,8 @@ define Device/Default
   DEVICE_DTS = freescale/$(subst _,-,$(1))
   SUPPORTED_DEVICES = $(subst _,$(comma),$(1))
   IMAGE_SIZE := 64m
-  IMAGE/sysupgrade.bin := append-kernel | pad-to 16M | \
+  IMAGE/sysupgrade.bin = ls-append-dtb $$(DEVICE_DTS) | pad-to 1M | \
+	append-kernel | pad-to 17M | \
```
Comment (adschm) [inline target/linux/layerscape/image/armv8_64b.mk:19]: Indent should be spaces here.

File: target/linux/layerscape/image/armv7.mk (hunk 7 above)
Comment (CHKDSK88) [reply target/linux/layerscape/image/armv7.mk]: Done.

**Comment by CHKDSK88** [comment]
@yangbolu1991 

Could You look at this PR?

**Comment by CHKDSK88** [comment]
@adschm 

Should I create `base-files/etc/board.d/05_compat-version` instead `uci-defaults/05_fix-compat-version`? Or both?

**Comment by adschm** [comment]
@CHKDSK88 For this PR the situation is different:
`base-files/etc/board.d/05_compat-version` will only bump the config-version if _config is removed_, i.e. with sysupgrade -n. Since in your case that's not desired, it wouldn't work with just sysupgrade -F and the old config would still be kept. That why you need uci-defaults/05_fix-compat-version since that changes the config-version on _every_ upgrade, not just if the config is wiped. See also my extensive explanation here: https://github.com/openwrt/openwrt/pull/3345#issuecomment-680143028

So, I think your current solution is appropriate.

File: target/linux/layerscape/base-files/lib/upgrade/platform.sh [hunk 9]
```diff
@@ -60,7 +60,6 @@ platform_do_upgrade_sdboot() {
 
 	if [ -n "$diff" ]; then
 		dd if="$1" of="/dev/$diskdev" bs=1024 count=4 > /dev/null 2>&1
-		dd if="$1" of="$diskdev" bs=1024 skip=4 seek=16384 > /dev/null 2>&1
```
Comment (yangbolu1991) [inline target/linux/layerscape/base-files/lib/upgrade/platform.sh:63]: This line should be fixed as below instead of removing. The line is writing kernel+rootfs for non-partition-saving case.
dd if="$1" of="/dev/$diskdev" bs=1024 skip=4 seek=16384 > /dev/null 2>&1

File: target/linux/layerscape/base-files/lib/upgrade/platform.sh [hunk 10]
```diff
@@ -70,6 +70,9 @@ platform_do_upgrade_sdboot() {
 		return 0
 	fi
 
+	# write dtb
+	echo "Writing dtb to /dev/$diskdev..."
+	dd if="$1" of="/dev/$diskdev" bs=1024 skip=15360 seek=15360 count=1024 > /dev/null 2>&1
```
Comment (yangbolu1991) [inline target/linux/layerscape/base-files/lib/upgrade/platform.sh]: The dtb writting should be added to if [ -n "$diff" ] section too.

File: package/boot/uboot-layerscape/files/fsl_ls1012a-frwy-sdboot-uEnv.txt [hunk 11]
```diff
@@ -1,8 +1,8 @@
 fdtaddr=0x8f000000
 loadaddr=0x81000000
-fdt_high=0xffffffffffffffff
+fdt_high=0x8fffffff
 initrd_high=0xffffffffffffffff
-sd_boot=mmc read $fdtaddr 7800 800;mmc read $loadaddr 8000 8000;bootm $loadaddr - $fdtaddr
-bootargs=root=/dev/mmcblk0p1 rw rootwait rootfstype=ext4 noinitrd earlycon=uart8250,mmio,0x21c0500 console=ttyS0,115200
+sd_boot=ext4load mmc 0:1 $loadaddr fitImage;bootm $loadaddr
+bootargs=root=/dev/mmcblk0p2 rw rootwait rootfstype=squashfs,ext4 noinitrd earlycon=uart8250,mmio,0x21c0500 console=ttyS0,115200
```
Comment (yangbolu1991) [inline package/boot/uboot-layerscape/files/fsl_ls1012a-frwy-sdboot-uEnv.txt:6]: To use overlay, the kmod-fs-f2fs and mkf2fs should be enabled in default, right?

**Comment by yangbolu1991** [comment]
@CHKDSK88 
Added some minor comments.  Beside these, all patches looks good to me.
And I appreciate for your patches on layerscape.
Thanks.

**Comment by adschm** [comment]
@yangbolu1991 Thanks for reviewing. Would you provide a Reviewed-by then?

Is your review complete enough so that I can merge this PR based on your judgement (as I'm not familiar enough with the details of the layerscape platform)?

**Comment by yangbolu1991** [comment]
@adschm 
Yes. I had went through all patches, and done a basic sysupgrade test on ls1043ardb sd card boot (with kmod-fs-f2fs and mkf2fs enabled). It seemed fine.
With comments addressed,
Reviewed-by: Yangbo Lu <yangbo.lu@nxp.com>

Thanks.

File: package/boot/uboot-layerscape/files/fsl_ls1012a-frwy-sdboot-uEnv.txt (hunk 11 above)
Comment (CHKDSK88) [reply package/boot/uboot-layerscape/files/fsl_ls1012a-frwy-sdboot-uEnv.txt:6]: I use ext4 overlay like other targets. No need kmod-fs-f2fs and mkf2fs.

Comment (yangbolu1991) [reply package/boot/uboot-layerscape/files/fsl_ls1012a-frwy-sdboot-uEnv.txt:6]: I see... Thanks.
It seemed to be some format issue after squashfs. I will check too.

[    6.912131] mount_root: rootdisk overlay filesystem has not been formatted yet
sh: [    6.920116] mount_root: no usable overlay filesystem found, using tmpfs overlay

Comment (CHKDSK88) [reply package/boot/uboot-layerscape/files/fsl_ls1012a-frwy-sdboot-uEnv.txt:6]: It should be:

```
[    6.787073] mount_root: rootdisk overlay filesystem has not been formatted yet
[    7.013550] random: mkfs.ext4: uninitialized urandom read (16 bytes read)
[    7.067903] random: fast init done
[   11.519161] EXT4-fs (loop0): mounted filesystem with ordered data mode. Opts: (null)
[   11.526922] ext4 filesystem being mounted at /tmp/overlay supports timestamps until 2038 (0x7fffffff)
[   11.536265] mount_root: overlay filesystem has not been fully initialized yet
[   11.543684] mount_root: switching to ext4 overlay
```
E2fsprogs should be default package:
https://github.com/openwrt/openwrt/blob/0b555f683efa63d9f4b3a9d64f8f028a48d07386/target/linux/layerscape/Makefile#L22
Do You have e2fsprogs added to image?

File: target/linux/layerscape/base-files/lib/upgrade/platform.sh (hunk 9 above)
Comment (CHKDSK88) [reply target/linux/layerscape/base-files/lib/upgrade/platform.sh:63]: It shouldn't be here. In this moment kernel is written on every case:
https://github.com/openwrt/openwrt/blob/ffc3a1ea8bc0e9306f05bd08f62fd80e5e732a07/target/linux/layerscape/base-files/lib/upgrade/platform.sh#L75

If we keep this line, kernel will be written two times for non-partition-saving case.

**Comment by adschm** [comment]
I've picked the first two patches and the last one for now, as those have no open points to be discussed.

**Comment by CHKDSK88** [comment]
@adschm 

Thank You. I rebased my PR.

@yangbolu1991 
Are You ok with my explanations?

File: target/linux/layerscape/base-files/lib/upgrade/platform.sh (hunk 9 above)
Comment (yangbolu1991) [reply target/linux/layerscape/base-files/lib/upgrade/platform.sh:63]: No. There is a return in "if [ -n "$diff" ]" section.

File: package/boot/uboot-layerscape/files/fsl_ls1012a-frwy-sdboot-uEnv.txt (hunk 11 above)
Comment (yangbolu1991) [reply package/boot/uboot-layerscape/files/fsl_ls1012a-frwy-sdboot-uEnv.txt:6]: Yes. I double checked. Here is my full log. Not sure what I missed.

May I know which board you were verified on?
I built with below options enabled.
CONFIG_TARGET_MULTI_PROFILE=y
CONFIG_TARGET_PER_DEVICE_ROOTFS=y
CONFIG_TARGET_ALL_PROFILES=y

Thanks.

File: target/linux/layerscape/base-files/lib/upgrade/platform.sh (hunk 9 above)
Comment (CHKDSK88) [reply target/linux/layerscape/base-files/lib/upgrade/platform.sh:63]: Sorry, my bad.

File: package/boot/uboot-layerscape/files/fsl_ls1012a-frwy-sdboot-uEnv.txt (hunk 11 above)
Comment (CHKDSK88) [reply package/boot/uboot-layerscape/files/fsl_ls1012a-frwy-sdboot-uEnv.txt]: @yangbolu1991 

My board is LS1046ARDB.
I used "make distclean", wiped my whole SD and I reproduced You state.
I reworked last commit and added missing f2fs features. Could You check if everything is ok?

Comment (yangbolu1991) [reply package/boot/uboot-layerscape/files/fsl_ls1012a-frwy-sdboot-uEnv.txt]: Now it's ok. It seems f2fs is used for root partition larger than 100MB, and ext4 is used for root partiton lower than 100MB.

**Comment by yangbolu1991** [comment]
@CHKDSK88 @adschm 
Current patches look good to me.
Thanks a lot.

**Comment by CHKDSK88** [comment]
@adschm 

Ready for merge?

File: package/boot/uboot-layerscape/files/fsl_ls1012a-frwy-sdboot-uEnv.txt (hunk 11 above)
Comment (CHKDSK88) [reply package/boot/uboot-layerscape/files/fsl_ls1012a-frwy-sdboot-uEnv.txt]: @yangbolu1991 

Thank You,

File: target/linux/layerscape/base-files/etc/uci-defaults/05_fix-compat-version [hunk 12]
```diff
@@ -5,11 +5,16 @@
 . /lib/functions.sh
 
 case "$(board_name)" in
+	fsl,ls1012a-frwy-sdboot | \
 	fsl,ls1012a-rdb | \
```
Comment (adschm) [inline target/linux/layerscape/base-files/etc/uci-defaults/05_fix-compat-version:9]: `fsl,ls1021a-iot-sdboot | \` should be added here as well.

File: target/linux/layerscape/image/armv7.mk [hunk 13]
```diff
@@ -40,50 +40,41 @@ endef
...
   DEVICE_VARIANT := SD Card Boot
   DEVICE_DTS := ls1021a-twr
-  FILESYSTEMS := ext4
-  IMAGES := sdcard.img sysupgrade.bin
-  IMAGE/sdcard.img := \
-    ls-clean | \
-    ls-append-sdhead $(1) | pad-to 4K | \
-    ls-append $(1)-uboot.bin | pad-to 3M | \
-    ls-append $(1)-uboot-env.bin | pad-to 15M | \
-    ls-append-dtb $$(DEVICE_DTS) | pad-to 16M | \
```
Comment (adschm) [inline target/linux/layerscape/image/armv7.mk:52]: By removing this command, we silently drop 
`$(call Image/BuildDTB,$(DTS_DIR)/$(1).dts,$(DTS_DIR)/$(1).dtb)`
from Build/ls-append-dtb

Therefore, build fails with missing dtb for these devices:

FATAL ERROR: Couldn't open "/data/openwrt/build_dir/target-aarch64_generic_musl/linux-layerscape_armv8_64b/linux-5.4.71/arch/arm64/boot/dts/freescale/fsl-ls1012a-frwy-sdboot.dtb": No such file or directory

File: target/linux/layerscape/image/Makefile [hunk 14]
```diff
@@ -27,8 +29,16 @@ define Build/ls-append-dtb
 	dd if=$(DTS_DIR)/$(1).dtb >> $@
 endef
 
+define Build/ls-append-kernel
+	mkdir -p $(STAGING_DIR)/kerneldir-tmp
```
Comment (adschm) [inline target/linux/layerscape/image/Makefile:33]: Will this directory be unique for each device?

I'd rather use something like
```
	mkdir -p $@.tmp
	cp $(IMAGE_KERNEL) $@.tmp/fitImage
	make_ext4fs -J -L kernel -l "$(LS_SD_KERNELPART_SIZE)M" "$@.kernel.part" "$@.tmp"
	dd if=$@.kernel.part >> $@
```

where the names are constructed from the "image name" $@, and thus inherently unique.

The idea is prototyped here:
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/image/generic.mk#L56

Comment (adschm) [reply target/linux/layerscape/image/Makefile:33]: BTW: The intermediate files should also be removed again.

File: target/linux/layerscape/image/armv7.mk [hunk 15]
```diff
@@ -40,50 +40,41 @@ endef
...
 
 define Device/fsl_ls1021a-iot-sdboot
+  $(Device/rework-sdcard-images)
   DEVICE_VENDOR := NXP
   DEVICE_MODEL := LS1021A-IoT
   DEVICE_VARIANT := SD Card Boot
   DEVICE_DTS := ls1021a-iot
-  FILESYSTEMS := ext4
   SUPPORTED_DEVICES :=
-  IMAGES := sdcard.img sysupgrade.bin
```
Comment (adschm) [inline target/linux/layerscape/image/armv7.mk]: For this device, IMAGES is removed but not added back.

File: target/linux/layerscape/image/gen_sdcard_head_img.sh [hunk 16]
```diff
@@ -7,16 +7,18 @@
 #
 
 set -x
-[ $# -eq 3 ] || {
-    echo "SYNTAX: $0 <file> <rootfs part offset> <rootfs size>"
+[ $# -eq 5 ] || {
+    echo "SYNTAX: $0 <file> <kernel part offset> <kernel size> <kernel part offset> <rootfs size>"
```
Comment (adschm) [inline target/linux/layerscape/image/gen_sdcard_head_img.sh:11]: fourth argument should be "rootfs part offset"

File: target/linux/layerscape/image/armv7.mk (hunk 13 above)
Comment (CHKDSK88) [reply target/linux/layerscape/image/armv7.mk:52]: Now I used `Image/Prepare` like [archs38](https://github.com/openwrt/openwrt/blob/master/target/linux/archs38/image/Makefile):
```
define Image/Prepare
	# Build .dtb for all boards we may run on
	$(foreach dts,$(DEVICE_DTS_LIST),
		$(call Image/BuildDTB,$(DTS_DIR)/$(dts).dts,$(DTS_DIR)/$(dts).dtb)
	)
endef
```

I run clean build for fsl-ls1012a-frwy-sdboot and it works.

File: target/linux/layerscape/image/Makefile (hunk 14 above)
Comment (CHKDSK88) [reply target/linux/layerscape/image/Makefile]: This is better approach. I reworked it.

File: target/linux/layerscape/image/armv7.mk (hunk 15 above)
Comment (CHKDSK88) [reply target/linux/layerscape/image/armv7.mk]: Done.

File: target/linux/layerscape/image/gen_sdcard_head_img.sh (hunk 16 above)
Comment (CHKDSK88) [reply target/linux/layerscape/image/gen_sdcard_head_img.sh]: Done.

**Comment by CHKDSK88** [comment]
@adschm 
All requested changes done.

Will this PR be merged before #3521

File: target/linux/layerscape/image/armv7.mk [hunk 17]
```diff
@@ -40,50 +40,40 @@ endef
...
-    ls-append-sdhead $(1) | pad-to 15M | \
-    ls-append-dtb $$(DEVICE_DTS) | pad-to 16M | \
-    append-kernel | pad-to $(LS_SD_ROOTFSPART_OFFSET)M | \
-    append-rootfs | check-size $(LS_SD_IMAGE_SIZE) | append-metadata
+  KERNEL := kernel-bin | gzip | fit gzip $$(DTS_DIR)/$$(DEVICE_DTS).dtb
+  IMAGES := sdcard.img.gz sysupgrade.bin
+  IMAGE/sdcard.img.gz := \
+	ls-clean | \
+	ls-append-sdhead $(1) | pad-to 4K | \
+	ls-append $(1)-uboot.bin | pad-to 3M | \
```
Comment (adschm) [inline target/linux/layerscape/image/armv7.mk:73]: Are the changes for these two devices intended? They haven't been in the previous version and I don't see why the position of uboot-env should have moved?

File: target/linux/layerscape/image/Makefile (hunk 14 above)
Comment (adschm) [reply target/linux/layerscape/image/Makefile]: I think we should remove $@.kernel.part as well?
And probably it would be good to chain the commands like in the reference, i.e. add ` && \` after each line?

File: target/linux/layerscape/image/armv7.mk (hunk 17 above)
Comment (CHKDSK88) [reply target/linux/layerscape/image/armv7.mk]: Thank You, miss-copy from another commit.

File: target/linux/layerscape/image/Makefile (hunk 14 above)
Comment (CHKDSK88) [reply target/linux/layerscape/image/Makefile]: Done.

**Comment by yangbolu1991** [comment]
> 
> 
> @adschm
> All requested changes done.
> 
> Will this PR be merged before #3521 ?

Actually I prefer this pull request merged first.
It had been created for a long time. It fixes sysupgrade bugs, so that my pull request adding new boards could work.
I will rebase my pull request if this one is merged.
Thanks.

**Comment by adschm** [comment]
I plan to merge this one soon, at the moment I'm stuck with strange build failure or armv7, where I'm not able to build, but don't see what's actually breaking. armv8_64b built fine in my test yesterday now.

**Comment by CHKDSK88** [comment]
@adschm 

What error do You have?

**Comment by adschm** [comment]
I don't have so much time this weekend. The problem is that I actually have problems find where it fails.

**Comment by adschm** [comment]
Seems to be the following, however I don't see how we touch that compared to the build bots:

```
make[4]: Entering directory '/data/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/restool-LSDK-20.04'
arm-openwrt-linux-muslgnueabi-gcc  -Wall -Wstrict-prototypes -Wextra -Wformat -std=gnu99 -Wmissing-prototypes -Wpointer-arith -Winline -Werror -Wundef -Icommon/ -DVERSION=\"v1.6\" -c mc_v10/dprc.c -o mc_v10/dprc.o
cc1: note: someone does not honour COPTS correctly, passed 0 times
In file included from mc_v10/dprc.c:34:
mc_v10/dprc.c: In function 'dprc_get_res_ids':
mc_v10/fsl_mc_cmd.h:56:28: error: right shift count >= width of type [-Werror=shift-count-overflow]
   (((~0UL) << (l)) & (~0UL >> (BITS_PER_LONG - 1 - (h))))
                            ^~
mc_v10/fsl_dprc_cmd.h:77:2: note: in expansion of macro 'GENMASK'
  GENMASK(DPRC_##field##_SHIFT + DPRC_##field##_SIZE - 1, \
  ^~~~~~~
mc_v10/fsl_dprc_cmd.h:80:47: note: in expansion of macro 'DPRC_MASK'
  ((var) |= (((val) << DPRC_##field##_SHIFT) & DPRC_MASK(field)))
                                               ^~~~~~~~~
mc_v10/dprc.c:611:2: note: in expansion of macro 'dprc_set_field'
  dprc_set_field(cmd_params->iter_status_lo,
  ^~~~~~~~~~~~~~
mc_v10/fsl_mc_cmd.h:56:28: error: right shift count >= width of type [-Werror=shift-count-overflow]
   (((~0UL) << (l)) & (~0UL >> (BITS_PER_LONG - 1 - (h))))
                            ^~
mc_v10/fsl_dprc_cmd.h:77:2: note: in expansion of macro 'GENMASK'
  GENMASK(DPRC_##field##_SHIFT + DPRC_##field##_SIZE - 1, \
  ^~~~~~~
mc_v10/fsl_dprc_cmd.h:80:47: note: in expansion of macro 'DPRC_MASK'
  ((var) |= (((val) << DPRC_##field##_SHIFT) & DPRC_MASK(field)))
                                               ^~~~~~~~~
mc_v10/dprc.c:614:2: note: in expansion of macro 'dprc_set_field'
  dprc_set_field(cmd_params->iter_status_hi,
  ^~~~~~~~~~~~~~
mc_v10/fsl_mc_cmd.h:56:28: error: right shift count >= width of type [-Werror=shift-count-overflow]
   (((~0UL) << (l)) & (~0UL >> (BITS_PER_LONG - 1 - (h))))
                            ^~
mc_v10/fsl_dprc_cmd.h:77:2: note: in expansion of macro 'GENMASK'
  GENMASK(DPRC_##field##_SHIFT + DPRC_##field##_SIZE - 1, \
  ^~~~~~~
mc_v10/fsl_dprc_cmd.h:82:12: note: in expansion of macro 'DPRC_MASK'
  (((var) & DPRC_MASK(field)) >> DPRC_##field##_SHIFT)
            ^~~~~~~~~
mc_v10/dprc.c:627:12: note: in expansion of macro 'dprc_get_field'
  iter_lo = dprc_get_field(rsp_params->iter_status_lo, ITER_STATUS_LO);
            ^~~~~~~~~~~~~~
mc_v10/fsl_mc_cmd.h:56:28: error: right shift count >= width of type [-Werror=shift-count-overflow]
   (((~0UL) << (l)) & (~0UL >> (BITS_PER_LONG - 1 - (h))))
                            ^~
mc_v10/fsl_dprc_cmd.h:77:2: note: in expansion of macro 'GENMASK'
  GENMASK(DPRC_##field##_SHIFT + DPRC_##field##_SIZE - 1, \
  ^~~~~~~
mc_v10/fsl_dprc_cmd.h:82:12: note: in expansion of macro 'DPRC_MASK'
  (((var) & DPRC_MASK(field)) >> DPRC_##field##_SHIFT)
            ^~~~~~~~~
mc_v10/dprc.c:628:12: note: in expansion of macro 'dprc_get_field'
  iter_hi = dprc_get_field(rsp_params->iter_status_hi, ITER_STATUS_HI);
            ^~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[4]: *** [Makefile:40: mc_v10/dprc.o] Error 1
make[4]: Leaving directory '/data/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/restool-LSDK-20.04'
make[3]: *** [Makefile:40: /data/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/restool-LSDK-20.04/.built] Error 2
make[3]: Leaving directory '/data/openwrt/package/network/utils/layerscape/restool'
time: package/network/utils/layerscape/restool/compile#0.20#0.03#0.33
make[2]: *** [package/Makefile:114: package/network/utils/layerscape/restool/compile] Error 2
make[2]: Leaving directory '/data/openwrt'
make[1]: *** [package/Makefile:108: /data/openwrt/staging_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/stamp/.package_compile] Error 2
make[1]: Leaving directory '/data/openwrt'
make: *** [/data/openwrt/include/toplevel.mk:236: world] Error 2
```

I was using buildbot settings on armv7.

File: target/linux/layerscape/image/armv7.mk [hunk 18]
```diff
@@ -40,50 +40,40 @@ endef
...
-    append-kernel | pad-to $(LS_SD_ROOTFSPART_OFFSET)M | \
-    append-rootfs | check-size $(LS_SD_IMAGE_SIZE) | append-metadata
+  KERNEL := kernel-bin | gzip | fit gzip $$(DTS_DIR)/$$(DEVICE_DTS).dtb
+  IMAGES := sdcard.img.gz sysupgrade.bin
+  IMAGE/sdcard.img.gz := \
+	ls-clean | \
+	ls-append-sdhead $(1) | pad-to 4K | \
+	ls-append $(1)-uboot.bin | pad-to 1M | \
+	ls-append $(1)-uboot-env.bin | pad-to 16M | \
+	ls-append-kernel | pad-to $(LS_SD_ROOTFSPART_OFFSET)M | \
```
Comment (adschm) [inline target/linux/layerscape/image/armv7.mk:75]: Why is check-size replaced by pad-to here? Won't that create additional useless writes to the SD card?

**Comment by adschm** [comment]
Thanks, I've merged this one now.

@CHKDSK88 I'd still be interested in feedback regarding the pad-to.
If we don't need it, we can still go back to check-size or just drop it in a separate patch.

**Comment by CHKDSK88** [comment]
@adschm 

Thank You.

Please look:
https://github.com/openwrt/openwrt/blob/a4c1c169c391ac0b11e4c3e895b282f07585347c/scripts/gen_image_generic.sh#L29
I follow this method and fill all rootfs with zeros.
This prevent from trash stored on unflushed sd-card.

**Comment by yangbolu1991** [comment]
> 
> 
> I plan to merge this one soon, at the moment I'm stuck with strange build failure on armv7, where I'm not able to build, but don't see what's actually breaking. armv8_64b built fine in my test yesterday now.

The restool package is for Layerscape DPAA2 platforms which are armv8 64b platforms. It seems it doesn't consider build for armv7. Let me add layerscape_armv8_64b dependency for restool package.
Thanks.

**Comment by adschm** [comment]
> > I plan to merge this one soon, at the moment I'm stuck with strange build failure on armv7, where I'm not able to build, but don't see what's actually breaking. armv8_64b built fine in my test yesterday now.
> 
> The restool package is for Layerscape DPAA2 platforms which are armv8 64b platforms. It seems it doesn't consider build for armv7. Let me add layerscape_armv8_64b dependency for restool package.
> Thanks.

That might explain it. The strangeness was that it seems to work for buildbots and just fails for my local build with buildbot settings (i.e. seems to be an SDK vs. non-SDK issue). But if we can "remove" it anyway, that's even better.

**Comment by adschm** [comment]
> @adschm
> 
> Thank You.
> 
> Please look:
> https://github.com/openwrt/openwrt/blob/a4c1c169c391ac0b11e4c3e895b282f07585347c/scripts/gen_image_generic.sh#L29
> 
> 
> I follow this method and fill rest of rootfs with zeros.
> This prevent from trash stored on unflushed sd-card.

Okay, I understand. In other situations, people tended to prevent writing zeros in order to extend the lifetime of the flash written to, so I wanted to make sure ...
